### DISCUSSION TITLE
CPMG fitting fails with "ValueError: Calculation stopped"

### DISCUSSION THREAD

**Discussion by mrag-work** [gbouvignies/ChemEx#257] [Q&A] [answered]
Greetings,

I am attempting to fit CPMG data. ChemEx seems to have no issues reading the data but cannot fit the data. I do not believe this is  related to parameters (initial values, etc.) of fitting. 

```
Loading datasets...
  • Reading experiments/600mhz_expt.toml (cpmg_ch3_13c_h2c) -> 2 profiles

Reading default parameters...

Reading methods...

Starting the fits...
                                                   
 ────────────── Running STEP1 (1/2)  ──────────────
                                                   
  • Fit method -> leastsq
  • Setting parameter status...
                                
     Name     Status   Updated  
    ─────────────────────────── 
     KEX_AB   fixed          1  
     PB       fixed          1  
                                
  • Running the minimization...
                                                 
   ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅ Group 1_A1CD1 (1/2) ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅

     Iteration            χ²    Reduced χ²
    ───────────────────────────────────────

 -- Got a ValueError: Calculation stopped --
```

ChemEx version: 2024.5.3

I have attached the entire chemex directory here for troubleshooting. Any pointers is much appreciated. 

Thank you very much!

[chemex_for_github.tar.gz](https://github.com/user-attachments/files/16364195/chemex_for_github.tar.gz)

**Comment by gbouvignies** [comment]
Hello @mrag-work,

The issue appears to be related to missing chemical shift information for the resonances you're analyzing. In the `parameters.toml` file, the `[CS_A]` section is missing crucial data. This section should contain the chemical shifts of the A state for each resonance, which are necessary for accurate offset calculations relative to the carrier frequency.

To resolve this, please add the following to your `parameters.toml` file:

```toml
[CS_A]
A1CD1 = 12.255
A2CD1 = 13.745
```

Replace these values with the actual chemical shifts of your resonances.

After making this change, try running the fit again. This should provide ChemEx with the necessary information to properly calculate the offsets and proceed with the fitting process.

If you continue to experience issues after making this change, please don't hesitate to reach out with the updated results or any new error messages.

Let me know if you need any further clarification or assistance!

**Comment by mrag-work** [comment]
Hello @gbouvignies ,

Thank you for your reply.

My apologies. I had the wrong file in the uploaded tarball. Here is the correct ```parameters.toml```.

```
% cat parameters/parameters.toml 
[GLOBAL]
PB      =   0.05
KEX_AB  =   1500
DW_AB   =   0.2
R1_A    =   1.5
R2_A    =   10.0
R1A_A   =   3.0

[CS_A]
A1CD1 = 12.255
A2CD1 = 13.745
```

Here is the result of the fitting - 

```
% ./run.sh 
╭──────────────────────────────────────────╮
│                                          │
│    ________                   ______     │
│   / ____/ /_  ___  ____ ___  / ____/  __ │
│  / /   / __ \/ _ \/ __ `__ \/ __/ | |/_/ │
│ / /___/ / / /  __/ / / / / / /____>  <   │
│ \____/_/ /_/\___/_/ /_/ /_/_____/_/|_|   │
│                                          │
│                                          │
│ Analysis of NMR chemical exchange data   │
│                                          │
│ Version: 2024.5.3                        │
╰──────────────────────────────────────────╯

Loading datasets...
  • Reading experiments/600mhz_expt.toml (cpmg_ch3_13c_h2c) -> 2 profiles

Reading default parameters...

Reading methods...

Starting the fits...
                                                   
 ────────────── Running STEP1 (1/2)  ──────────────
                                                   
  • Fit method -> leastsq
  • Setting parameter status...
                                
     Name     Status   Updated  
    ─────────────────────────── 
     PB       fixed          1  
     KEX_AB   fixed          1  
                                
  • Running the minimization...
                                                 
   ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅ Group 1_A1CD1 (1/2) ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅

     Iteration            χ²    Reduced χ²
    ───────────────────────────────────────

 -- Got a ValueError: Calculation stopped --

  • Writing results in output/STEP1/Groups/1_A1CD1
                                                 
   ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅ Group 2_A2CD1 (2/2) ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅

     Iteration            χ²    Reduced χ²
    ───────────────────────────────────────

 -- Got a ValueError: Calculation stopped --

  • Writing results in output/STEP1/Groups/2_A2CD1
                                                 
   ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅ All groups ⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅

           χ²:             inf
   Reduced χ²:             inf

  • Writing results in output/STEP1/All
  • Making plots...
    ‣ output/STEP1/All/Plots/600mhz_expt.pdf [.fit, .exp]
                                                   
 ────────────── Running STEP2 (2/2)  ──────────────
                                                   
  • Fit method -> leastsq
  • Setting parameter status...
                                
     Name     Status   Updated  
    ─────────────────────────── 
     PB       fitted         1  
     KEX_AB   fitted         1  
                                
  • Running the minimization...

     Iteration            χ²    Reduced χ²
    ───────────────────────────────────────

 -- Got a ValueError: Calculation stopped --

  • Writing results in output/STEP2
  • Making plots...
    ‣ output/STEP2/Plots/600mhz_expt.pdf [.fit, .exp]
```

**Comment by gbouvignies** [comment, marked as answer]
Hello @mrag-work,

I think I've identified the issue and have a solution for you. The problem stems from your profiles containing duplicate points with identical values, which isn't typically expected. Here's what's happening:

1. You've set `error = "duplicates"` in your experiment configuration.
2. With identical duplicate values, the calculated uncertainty becomes zero.
3. This causes a division by zero when computing the chi-square value, leading to the fit failure.

To fix this:

1. Open your `experiments.toml` file.
2. Find the `error` field in the `[data]` section.
3. Change it from `"duplicates"` to `"file"`:

```toml
[data]
path  = "../data/600MHz/"
error = "file"
```

This will make ChemEx use the error values from your data file instead of calculating them from duplicates.

Try running the fit again with this change. It should work without the ValueError.

Thank you for bringing this to my attention. I'll look into adding checks for this situation in future ChemEx versions.

If you have any questions or encounter further issues, please let me know. Good luck with your analysis!

**Comment by mrag-work** [comment]
Hi @gbouvignies 

Changing the ```error``` field from ```duplicates``` to ```file``` fixed the issue!

Thank you very much! Much appreciated. I have marked your response as the answer.
